ocaml/xapi/xapi_pif.ml

```diff
@@ -753,15 +757,28 @@ let reconfigure_ipv6 ~__context ~self ~mode ~iPv6 ~gateway ~dNS =
   Db.PIF.set_ipv6_configuration_mode ~__context ~self ~value:mode ;
   Db.PIF.set_ipv6_gateway ~__context ~self ~value:gateway ;
   Db.PIF.set_IPv6 ~__context ~self ~value:[iPv6] ;
-  (* Keep IPv4 DNS entries *)
-  let pif_dns =
+  (* Only keep existing IPv4 entries *)
+  let pif_ipv4_dns =
     List.filter
       (fun dns_addr ->
         Xapi_stdext_unix.Unixext.domain_of_addr dns_addr = Some Unix.PF_INET
       )
-      (String.split_on_char ',' (Db.PIF.get_DNS ~__context ~self))
+      (Astring.String.cuts ~sep:"," ~empty:false
+         (Db.PIF.get_DNS ~__context ~self)
+      )
+  in
+  let new_dns =
+    match (ipv6_dns_to_set, pif_ipv4_dns) with
+    | [], [] ->
+        ""
+    | _, [] ->
+        String.concat "," ipv6_dns_to_set
+    | [], _ ->
+        String.concat "," pif_ipv4_dns
+    | _, _ ->
+        String.concat "," (String.concat "," ipv6_dns_to_set :: pif_ipv4_dns)
```

Comment:
Both `ipv6_dns_to_set` and `pif_ipv4_dns` are lists,  `::` adds a single element to a list. Instead use `List.concat [ipv6; ipv4]` to concatenate the lists.